include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,808 @@
+   * use the erase() function.
+   *
+   * @note It is possible to use different types @p T for different cells which
+   * may reflect, for example, different constitutive models of continuum mechanics
+   * in different parts of the domain.
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell,
+                  const unsigned int number_of_data_points);
+
+  /**
+   * Same as above but for a range of iterators starting at @p cell_start
+   * until, but not including, @p cell_end for all locally owned cells, i.e.
+   * for which `is_locally_owned()==true`
+   */
+  template<typename T=DataType>
+  void initialize(const CellIteratorType &cell_start,
+                  const CellIteratorType &cell_end,
+                  const unsigned int number);
+
+  /**
+   * Removes data stored at the @p cell. Returns true if the data was removed.
+   * If no data is attached to the @p cell, this function will not do anything
+   * and returns false.
+   *
+   * @note In debug mode, this function will also check that the are no
+   * outstanding references to the data stored on this cell. That is to say,
+   * that the only references to the stored data are that made by this class.
+   */
+  bool erase(const CellIteratorType &cell);
+
+  /**
+   * Clear all the data stored in this object.
+   */
+  void clear();
+
+  /**
+   * Get a vector of the data located at @p cell .
+   * A possible additional typename @p T is the class to which the base class
+   * DataType could be cast. Since @p DataType is stored as shared pointers,
+   * there is minimal overhead in returning a vector by value instead of by reference.
+   * This allows flexibility if class @p T is not the same as @p DataType on a
+   * cell-by-cell basis.
+   */
+  template<typename T=DataType>
+  CellVector<T> get_data(const CellIteratorType &cell);
+
+  /**
+   * Get a vector of constant pointers to data located at @p cell .
+   * A possible additional typename @p T is the class to which the base class
+   * DataType could be cast. Since @p DataType is stored as shared pointers,
+   * there is minimal overhead in returning a vector by value instead of by reference.
+   * This allows flexibility if class @p T is not the same as @p DataType on a
+   * cell-by-cell basis.
+   */
+  template<typename T=DataType>
+  CellVector<const T> get_data(const CellIteratorType &cell) const;
+
+private:
+  /**
+   * A map to store a vector of data on a cell.
+   */
+  std::map<CellIteratorType, CellVector<DataType> > map;
+
+};
+
+
+/**
+ * An abstract class which specifies requirements for data on
+ * a single quadrature point to be transferable during refinement.
+ * One possible implementation is via the restriction/prolongation matrices
+ * as implemented in ContinuousQuadratureDataTransfer class.
+ *
+ * @author Denis Davydov, Jean-Paul Pelteret, 2016
+ */
+class TransferableQuadraturePointData
```

Comment:
Because we don't just ship this data from processor to processor. We first need to do L2 projection as described in the `ContinuousQuadratureDataTransfer` class. Thereby however you look at it, the data should be some typo of float. Doing L2 projection and then interpolation to new quadrature points of integer data does not make much sense anyway. As for complex-valued data, one can still pack/unpack it into doubles.